protocol/signalfx/trace.go

```diff
@@ -19,8 +22,430 @@ const (
+var (
+	trueVar = true
+
+	ClientKind   = "CLIENT"
+	ServerKind   = "SERVER"
+	ProducerKind = "PRODUCER"
+	ConsumerKind = "CONSUMER"
+)
+
+// asTraceSpan returns a slice of our SignalFx span object, which is equivalent
+// to a Zipkin V2 span.  The conversion code was mostly ported from
+// https://github.com/openzipkin/zipkin/blob/2.8.4/zipkin/src/main/java/zipkin/internal/V2SpanConverter.java
+func (is *inputSpan) asTraceSpans() ([]*trace.Span, error) {
+	// The presence of the "kind" field, tags, or local/remote endpoints is a
+	// dead giveaway that this is a Zipkin v2 span, so shortcut the whole
+	// process and return it as an optimization.  If it doesn't have any of
+	// those things it could still be a V2 span since none of them are strictly
+	// required to be there.
+	if is.Span.Kind != nil || len(is.Span.Tags) > 0 || is.Span.LocalEndpoint != nil || is.Span.RemoteEndpoint != nil {
+		// Do some basic validation
+		if len(is.BinaryAnnotations) > 0 {
+			return nil, errors.New("span cannot have binaryAnnotations with Zipkin V2 fields")
+		}
+
+		if len(is.Annotations) > 0 {
+			is.Span.Annotations = make([]*trace.Annotation, len(is.Annotations))
+			for i := range is.Annotations {
+				is.Span.Annotations[i] = is.Annotations[i].ToV2()
+			}
+		}
+		return []*trace.Span{&is.Span}, nil
+	}
+
+	// Otherwise if we can't prove it is a zipkin v2 span, we need to convert
+	// the input to a set of v2 spans.  A single span in Zipkin v1 can contain
+	// multiple v2 spans because the annotations and binary annotations contain
+	// endpoints.
+
+	if is.Span.Tags == nil {
+		is.Span.Tags = map[string]string{}
+	}
+
+	spanCopy := is.Span
+	spanBuilder := &spanBuilder{
+		spans: []*trace.Span{&spanCopy},
+	}
+	spanBuilder.processAnnotations(is)
+	if err := spanBuilder.processBinaryAnnotations(is); err != nil {
+		return nil, err
+	}
+
+	return spanBuilder.spans, nil
+}
+
+func (is *inputSpan) endTimestampReflectsSpanDuration(end *annotation) bool {
+	return end != nil && is.Timestamp != nil && is.Duration != nil && end.Timestamp != nil &&
+		*is.Timestamp+*is.Duration == *end.Timestamp
+}
+
+type spanBuilder struct {
+	spans                          []*trace.Span
+	cs, sr, ss, cr, ms, mr, ws, wr *annotation
+}
+
+func (sb *spanBuilder) addSpanForEndpoint(is *inputSpan, e *trace.Endpoint) *trace.Span {
+	s := is.Span
+	s.LocalEndpoint = e
+	s.Tags = map[string]string{}
+
+	sb.spans = append(sb.spans, &s)
+	return &s
+}
+
+func (sb *spanBuilder) spanForEndpoint(is *inputSpan, e *trace.Endpoint) *trace.Span {
+	if e == nil {
+		// Allocate missing endpoint data to first span.  For a Zipkin v2
+		// span this will be the only one.
+		return sb.spans[0]
+	}
+
+	for i := range sb.spans {
+		next := sb.spans[i]
+		if next.LocalEndpoint == nil {
+			next.LocalEndpoint = e
+			return next
+		} else if closeEnough(next.LocalEndpoint, e) {
+			return next
+		}
+	}
+
+	return sb.addSpanForEndpoint(is, e)
+}
+
+// nolint: gocyclo
```

Comment:
why can't you just fix it?

Reply:
I can, but I was just trying to make it as much a straight port of the Zipkin conversion code to Golang as possible so they can be easily compared.

Reply:
admirable, but it's a slippery slope.  Please fix for our linters.